// @angular-redux/store/esm2015/testing/ng-redux.mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TODO: See if this linting rule can be enabled with new build process (ng-packagr)
// tslint:disable:no-implicit-dependencies
// tslint:disable:member-ordering
import { NgRedux, } from '@angular-redux/store';
import { MockObservableStore } from './observable-store.mock';
/**
 * Convenience mock to make it easier to control selector
 * behaviour in unit tests.
 * @template T
 */
export class MockNgRedux extends NgRedux {
    /**
     * @hidden
     */
    constructor() {
        super();
        //
        this.mockRootStore = new MockObservableStore();
        this.configureSubStore = (/** @type {?} */ (this.mockRootStore.configureSubStore));
        this.dispatch = (/** @type {?} */ (this.mockRootStore.dispatch));
        this.getState = (/** @type {?} */ (this.mockRootStore.getState));
        this.subscribe = this.mockRootStore.subscribe;
        this.replaceReducer = this.mockRootStore.replaceReducer;
        this.select = this.mockRootStore.select;
        this.provideStore = (_) => undefined;
        this.configureStore = (_, __, ___, ____) => undefined;
        // This hooks the mock up to @select.
        NgRedux.instance = (/** @type {?} */ (this));
    }
    /**
     * Returns a subject that's connected to any observable returned by the
     * given selector. You can use this subject to pump values into your
     * components or services under test; when they call .select or \@select
     * in the context of a unit test, MockNgRedux will give them the values
     * you pushed onto your stub.
     * @template R, S
     * @param {?=} selector
     * @param {?=} comparator
     * @return {?}
     */
    static getSelectorStub(selector, comparator) {
        return MockNgRedux.getInstance().mockRootStore.getSelectorStub(selector, comparator);
    }
    /**
     * Returns a mock substore that allows you to set up selectorStubs for
     * any 'fractal' stores your app creates with NgRedux.configureSubStore.
     *
     * If your app creates deeply nested substores from other substores,
     * pass the chain of pathSelectors in as ordered arguments to mock
     * the nested substores out.
     * @template S
     * @param {...?} pathSelectors
     * @return {?}
     */
    static getSubStore(...pathSelectors) {
        return pathSelectors.length
            ? MockNgRedux.getInstance().mockRootStore.getSubStore(...pathSelectors)
            : MockNgRedux.getInstance().mockRootStore;
    }
    /**
     * Reset all previously configured stubs.
     * @return {?}
     */
    static reset() {
        MockNgRedux.getInstance().mockRootStore.reset();
        NgRedux.instance = (/** @type {?} */ (MockNgRedux.mockInstance));
    }
    /**
     * Gets the singleton MockNgRedux instance. Useful for cases where your
     * tests need to spy on store methods, for example.
     * @return {?}
     */
    static getInstance() {
        MockNgRedux.mockInstance = MockNgRedux.mockInstance || new MockNgRedux();
        return MockNgRedux.mockInstance;
    }
}
/**
 * @deprecated Use MockNgRedux.getInstance() instead.
 */
MockNgRedux.mockInstance = undefined;
if (false) {
    /**
     * @deprecated Use MockNgRedux.getInstance() instead.
     * @type {?}
     */
    MockNgRedux.mockInstance;
    /**
     * @type {?}
     * @private
     */
    MockNgRedux.prototype.mockRootStore;
    /** @type {?} */
    MockNgRedux.prototype.configureSubStore;
    /** @type {?} */
    MockNgRedux.prototype.dispatch;
    /** @type {?} */
    MockNgRedux.prototype.getState;
    /** @type {?} */
    MockNgRedux.prototype.subscribe;
    /** @type {?} */
    MockNgRedux.prototype.replaceReducer;
    /** @type {?} */
    MockNgRedux.prototype.select;
    /** @type {?} */
    MockNgRedux.prototype.provideStore;
    /** @type {?} */
    MockNgRedux.prototype.configureStore;
}
