// @angular-redux/store/esm2015/testing/observable-store.mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ReplaySubject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
/**
 * @hidden
 * @record
 */
export function SelectorStubRecord() { }
if (false) {
    /** @type {?} */
    SelectorStubRecord.prototype.subject;
    /** @type {?} */
    SelectorStubRecord.prototype.comparator;
}
/**
 * @hidden
 * @record
 */
export function SelectorStubMap() { }
/**
 * @hidden
 * @record
 */
export function SubStoreStubMap() { }
/**
 * @hidden
 * @template State
 */
export class MockObservableStore {
    constructor() {
        this.selections = {};
        this.subStores = {};
        this.getSelectorStub = (selector, comparator) => this.initSelectorStub(selector, comparator).subject;
        this.reset = () => {
            Object.keys(this.subStores).forEach(k => this.subStores[k].reset());
            this.selections = {};
            this.subStores = {};
        };
        this.dispatch = action => action;
        this.replaceReducer = () => null;
        this.getState = () => ({});
        this.subscribe = () => () => null;
        this.select = (selector, comparator) => {
            /** @type {?} */
            const stub = this.initSelectorStub(selector, comparator);
            return stub.comparator
                ? stub.subject.pipe(distinctUntilChanged(stub.comparator))
                : stub.subject;
        };
        this.configureSubStore = (basePath, _) => this.initSubStore(basePath);
        this.getSubStore = (...pathSelectors) => {
            const [first, ...rest] = pathSelectors;
            return (/** @type {?} */ ((first
                ? this.initSubStore(first).getSubStore(...rest)
                : this)));
        };
    }
    /**
     * @private
     * @template SubState
     * @param {?} basePath
     * @return {?}
     */
    initSubStore(basePath) {
        /** @type {?} */
        const result = this.subStores[JSON.stringify(basePath)] ||
            new MockObservableStore();
        this.subStores[JSON.stringify(basePath)] = result;
        return result;
    }
    /**
     * @private
     * @template SelectedState
     * @param {?=} selector
     * @param {?=} comparator
     * @return {?}
     */
    initSelectorStub(selector, comparator) {
        /** @type {?} */
        const key = selector ? selector.toString() : '';
        /** @type {?} */
        const record = this.selections[key] || {
            subject: new ReplaySubject(),
            comparator,
        };
        this.selections[key] = record;
        return record;
    }
}
if (false) {
    /** @type {?} */
    MockObservableStore.prototype.selections;
    /** @type {?} */
    MockObservableStore.prototype.subStores;
    /** @type {?} */
    MockObservableStore.prototype.getSelectorStub;
    /** @type {?} */
    MockObservableStore.prototype.reset;
    /** @type {?} */
    MockObservableStore.prototype.dispatch;
    /** @type {?} */
    MockObservableStore.prototype.replaceReducer;
    /** @type {?} */
    MockObservableStore.prototype.getState;
    /** @type {?} */
    MockObservableStore.prototype.subscribe;
    /** @type {?} */
    MockObservableStore.prototype.select;
    /** @type {?} */
    MockObservableStore.prototype.configureSubStore;
    /** @type {?} */
    MockObservableStore.prototype.getSubStore;
}
