// @angular-redux/store/esm5/testing/observable-store.mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ReplaySubject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
/**
 * @hidden
 * @record
 */
export function SelectorStubRecord() { }
if (false) {
    /** @type {?} */
    SelectorStubRecord.prototype.subject;
    /** @type {?} */
    SelectorStubRecord.prototype.comparator;
}
/**
 * @hidden
 * @record
 */
export function SelectorStubMap() { }
/**
 * @hidden
 * @record
 */
export function SubStoreStubMap() { }
/**
 * @hidden
 * @template State
 */
var /**
 * @hidden
 * @template State
 */
MockObservableStore = /** @class */ (function () {
    function MockObservableStore() {
        var _this = this;
        this.selections = {};
        this.subStores = {};
        this.getSelectorStub = function (selector, comparator) {
            return _this.initSelectorStub(selector, comparator).subject;
        };
        this.reset = function () {
            Object.keys(_this.subStores).forEach(function (k) { return _this.subStores[k].reset(); });
            _this.selections = {};
            _this.subStores = {};
        };
        this.dispatch = function (action) { return action; };
        this.replaceReducer = function () { return null; };
        this.getState = function () { return ({}); };
        this.subscribe = function () { return function () { return null; }; };
        this.select = function (selector, comparator) {
            /** @type {?} */
            var stub = _this.initSelectorStub(selector, comparator);
            return stub.comparator
                ? stub.subject.pipe(distinctUntilChanged(stub.comparator))
                : stub.subject;
        };
        this.configureSubStore = function (basePath, _) { return _this.initSubStore(basePath); };
        this.getSubStore = function () {
            var pathSelectors = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                pathSelectors[_i] = arguments[_i];
            }
            var _a;
            var _b = tslib_1.__read(pathSelectors), first = _b[0], rest = _b.slice(1);
            return (/** @type {?} */ ((first
                ? (_a = _this.initSubStore(first)).getSubStore.apply(_a, tslib_1.__spread(rest)) : _this)));
        };
    }
    /**
     * @private
     * @template SubState
     * @param {?} basePath
     * @return {?}
     */
    MockObservableStore.prototype.initSubStore = /**
     * @private
     * @template SubState
     * @param {?} basePath
     * @return {?}
     */
    function (basePath) {
        /** @type {?} */
        var result = this.subStores[JSON.stringify(basePath)] ||
            new MockObservableStore();
        this.subStores[JSON.stringify(basePath)] = result;
        return result;
    };
    /**
     * @private
     * @template SelectedState
     * @param {?=} selector
     * @param {?=} comparator
     * @return {?}
     */
    MockObservableStore.prototype.initSelectorStub = /**
     * @private
     * @template SelectedState
     * @param {?=} selector
     * @param {?=} comparator
     * @return {?}
     */
    function (selector, comparator) {
        /** @type {?} */
        var key = selector ? selector.toString() : '';
        /** @type {?} */
        var record = this.selections[key] || {
            subject: new ReplaySubject(),
            comparator: comparator,
        };
        this.selections[key] = record;
        return record;
    };
    return MockObservableStore;
}());
/**
 * @hidden
 * @template State
 */
export { MockObservableStore };
if (false) {
    /** @type {?} */
    MockObservableStore.prototype.selections;
    /** @type {?} */
    MockObservableStore.prototype.subStores;
    /** @type {?} */
    MockObservableStore.prototype.getSelectorStub;
    /** @type {?} */
    MockObservableStore.prototype.reset;
    /** @type {?} */
    MockObservableStore.prototype.dispatch;
    /** @type {?} */
    MockObservableStore.prototype.replaceReducer;
    /** @type {?} */
    MockObservableStore.prototype.getState;
    /** @type {?} */
    MockObservableStore.prototype.subscribe;
    /** @type {?} */
    MockObservableStore.prototype.select;
    /** @type {?} */
    MockObservableStore.prototype.configureSubStore;
    /** @type {?} */
    MockObservableStore.prototype.getSubStore;
}
