// @angular-redux/store/esm2015/components/sub-store.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { distinctUntilChanged, map } from 'rxjs/operators';
import { getIn } from '../utils/get-in';
import { registerFractalReducer, replaceLocalReducer, } from './fractal-reducer-map';
import { resolveToFunctionSelector, } from './selectors';
/**
 * @hidden
 * @template State
 */
export class SubStore {
    /**
     * @param {?} rootStore
     * @param {?} basePath
     * @param {?} localReducer
     */
    constructor(rootStore, basePath, localReducer) {
        this.rootStore = rootStore;
        this.basePath = basePath;
        this.dispatch = action => this.rootStore.dispatch(Object.assign({}, ((/** @type {?} */ (action))), { '@angular-redux::fractalkey': JSON.stringify(this.basePath) }));
        this.getState = () => getIn(this.rootStore.getState(), this.basePath);
        this.configureSubStore = (basePath, localReducer) => new SubStore(this.rootStore, [...this.basePath, ...basePath], localReducer);
        this.select = (selector, comparator) => this.rootStore.select(this.basePath).pipe(map(resolveToFunctionSelector(selector)), distinctUntilChanged(comparator));
        this.subscribe = (listener) => {
            /** @type {?} */
            const subscription = this.select().subscribe(listener);
            return () => subscription.unsubscribe();
        };
        this.replaceReducer = (nextLocalReducer) => replaceLocalReducer(this.basePath, nextLocalReducer);
        registerFractalReducer(basePath, localReducer);
    }
}
if (false) {
    /** @type {?} */
    SubStore.prototype.dispatch;
    /** @type {?} */
    SubStore.prototype.getState;
    /** @type {?} */
    SubStore.prototype.configureSubStore;
    /** @type {?} */
    SubStore.prototype.select;
    /** @type {?} */
    SubStore.prototype.subscribe;
    /** @type {?} */
    SubStore.prototype.replaceReducer;
    /**
     * @type {?}
     * @private
     */
    SubStore.prototype.rootStore;
    /**
     * @type {?}
     * @private
     */
    SubStore.prototype.basePath;
}
