// @angular-redux/store/esm5/components/dev-tools.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, Injectable, NgZone } from '@angular/core';
import { NgRedux } from './ng-redux';
/**
 * @record
 */
export function ReduxDevTools() { }
if (false) {
    /** @type {?} */
    ReduxDevTools.prototype.listen;
    /* Skipping unhandled member: (options: EnhancerOptions): StoreEnhancer<any>;*/
}
/**
 * @record
 */
function WindowWithReduxDevTools() { }
if (false) {
    /** @type {?|undefined} */
    WindowWithReduxDevTools.prototype.__REDUX_DEVTOOLS_EXTENSION__;
    /** @type {?|undefined} */
    WindowWithReduxDevTools.prototype.devToolsExtension;
}
/** @type {?} */
var environment = (/** @type {?} */ ((typeof window !== 'undefined'
    ? window
    : {})));
/**
 * An angular-2-ified version of the Redux DevTools chrome extension.
 */
var DevToolsExtension = /** @class */ (function () {
    /** @hidden */
    function DevToolsExtension(appRef, ngRedux) {
        var _this = this;
        this.appRef = appRef;
        this.ngRedux = ngRedux;
        /**
         * A wrapper for the Chrome Extension Redux DevTools.
         * Makes sure state changes triggered by the extension
         * trigger Angular2's change detector.
         *
         * @argument options: dev tool options; same
         * format as described here:
         * [zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md]
         */
        this.enhancer = function (options) {
            /** @type {?} */
            var subscription;
            if (!_this.isEnabled()) {
                return null;
            }
            // Make sure changes from dev tools update angular's view.
            (/** @type {?} */ (_this.getDevTools())).listen(function (_a) {
                var type = _a.type;
                if (type === 'START') {
                    subscription = _this.ngRedux.subscribe(function () {
                        if (!NgZone.isInAngularZone()) {
                            _this.appRef.tick();
                        }
                    });
                }
                else if (type === 'STOP') {
                    subscription();
                }
            });
            return (/** @type {?} */ (_this.getDevTools()))(options || {});
        };
        /**
         * Returns true if the extension is installed and enabled.
         */
        this.isEnabled = function () { return !!_this.getDevTools(); };
        /**
         * Returns the redux devtools enhancer.
         */
        this.getDevTools = function () {
            return environment &&
                (environment.__REDUX_DEVTOOLS_EXTENSION__ || environment.devToolsExtension);
        };
    }
    DevToolsExtension.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DevToolsExtension.ctorParameters = function () { return [
        { type: ApplicationRef },
        { type: NgRedux }
    ]; };
    return DevToolsExtension;
}());
export { DevToolsExtension };
if (false) {
    /**
     * A wrapper for the Chrome Extension Redux DevTools.
     * Makes sure state changes triggered by the extension
     * trigger Angular2's change detector.
     *
     * @argument options: dev tool options; same
     * format as described here:
     * [zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md]
     * @type {?}
     */
    DevToolsExtension.prototype.enhancer;
    /**
     * Returns true if the extension is installed and enabled.
     * @type {?}
     */
    DevToolsExtension.prototype.isEnabled;
    /**
     * Returns the redux devtools enhancer.
     * @type {?}
     */
    DevToolsExtension.prototype.getDevTools;
    /**
     * @type {?}
     * @private
     */
    DevToolsExtension.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    DevToolsExtension.prototype.ngRedux;
}
