// @angular-redux/store/esm5/components/selectors.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getIn } from '../utils/get-in';
/**
 * @hidden
 * @type {?}
 */
export var sniffSelectorType = function (selector) {
    return !selector
        ? 'nil'
        : Array.isArray(selector)
            ? 'path'
            : 'function' === typeof selector
                ? 'function'
                : 'property';
};
/**
 * @hidden
 * @type {?}
 */
export var resolver = function (selector) { return ({
    property: function (state) {
        return state ? state[(/** @type {?} */ (selector))] : undefined;
    },
    path: function (state) { return getIn(state, (/** @type {?} */ (selector))); },
    function: (/** @type {?} */ (selector)),
    nil: function (state) { return state; },
}); };
/**
 * @hidden
 * @type {?}
 */
export var resolveToFunctionSelector = function (selector) { return resolver(selector)[sniffSelectorType(selector)]; };
